pkg/gadgets/trace/bind/tracer/tracer_test.go

```diff
@@ -0,0 +1,649 @@
+	tracer.Stop()
+}
+
+type sockOpt struct {
+	level int
+	opt   int
+	value int
+}
+
+func TestBindTracer(t *testing.T) {
+	t.Parallel()
+
+	utilstest.RequireRoot(t)
+
+	const unprivilegedUID = int(1435)
+
+	type test struct {
+		getTracerConfig func(info *utilstest.RunnerInfo) *tracer.Config
+		runnerConfig    *utilstest.RunnerConfig
+		generateEvent   func() (uint16, error)
+		validateEvent   func(t *testing.T, info *utilstest.RunnerInfo, port uint16, events []types.Event)
+	}
+
+	for name, test := range map[string]test{
+		"captures_all_events_with_no_filters_configured": {
+			getTracerConfig: func(info *utilstest.RunnerInfo) *tracer.Config {
+				return &tracer.Config{}
+			},
+			generateEvent: bindSocketFn("127.0.0.1", unix.AF_INET, unix.SOCK_STREAM, 0),
+			validateEvent: utilstest.ExpectAtLeastOneEvent(func(info *utilstest.RunnerInfo, port uint16) *types.Event {
+				return &types.Event{
+					Event: eventtypes.Event{
+						Type: eventtypes.NORMAL,
+					},
+					Pid:       uint32(info.Pid),
+					Comm:      info.Comm,
+					Protocol:  "TCP",
+					Addr:      "127.0.0.1",
+					Port:      port,
+					Options:   ".....",
+					Interface: "",
+					MountNsID: info.MountNsID,
+				}
+			}),
+		},
+		"captures_no_events_with_no_matching_filter": {
+			getTracerConfig: func(info *utilstest.RunnerInfo) *tracer.Config {
+				return &tracer.Config{
+					MountnsMap: utilstest.CreateMntNsFilterMap(t, 0),
+				}
+			},
+			generateEvent: bindSocketFn("127.0.0.1", unix.AF_INET, unix.SOCK_STREAM, 0),
+			validateEvent: utilstest.ExpectNoEvent[types.Event, uint16],
+		},
+		"captures_events_with_matching_filter": {
+			getTracerConfig: func(info *utilstest.RunnerInfo) *tracer.Config {
+				return &tracer.Config{
+					MountnsMap: utilstest.CreateMntNsFilterMap(t, info.MountNsID),
+				}
+			},
+			generateEvent: bindSocketFn("127.0.0.1", unix.AF_INET, unix.SOCK_STREAM, 0),
+			validateEvent: utilstest.ExpectOneEvent(func(info *utilstest.RunnerInfo, port uint16) *types.Event {
+				return &types.Event{
+					Event: eventtypes.Event{
+						Type: eventtypes.NORMAL,
+					},
+					Pid:       uint32(info.Pid),
+					Comm:      info.Comm,
+					Protocol:  "TCP",
+					Addr:      "127.0.0.1",
+					Port:      port,
+					Options:   ".....",
+					Interface: "",
+					MountNsID: info.MountNsID,
+				}
+			}),
+		},
+		"ipv4": {
+			getTracerConfig: func(info *utilstest.RunnerInfo) *tracer.Config {
+				return &tracer.Config{
+					MountnsMap: utilstest.CreateMntNsFilterMap(t, info.MountNsID),
+				}
+			},
+			generateEvent: bindSocketFn("127.0.0.2", unix.AF_INET, unix.SOCK_STREAM, 0),
+			validateEvent: func(t *testing.T, info *utilstest.RunnerInfo, port uint16, events []types.Event) {
+				if len(events) != 1 {
+					t.Fatalf("Wrong number of events received %d, expected 1", len(events))
+				}
+
+				if events[0].Addr != "127.0.0.2" {
+					t.Fatalf("Event has wrong addr: %s, expected: %s",
+						events[0].Addr, "127.0.0.2")
+				}
+			},
```

Comment:
It's a very good question. The purpose of having test checking only one field is to keep them as focused in a single part of the system as possible. Otherwise will have a lot of tests that tests everything, and when something breaks it would be difficult to understand what as broken. IMO it's a good approach to have some tests that check all the fields, and then have other tests that check specific fields with more care. Probably @invidian has more thoughts on this.

Reply:
> Otherwise will have a lot of tests that tests everything, and when something breaks it would be difficult to understand what as broken.
> 
What would be difficult? You will have the name and value of the field that doesn't contain the expected result.

I think it all depends on how the fields are retrieved. For this specific case, the code for extracting the fields is the same except for the IP address. Therefore, it seems right to focus on a single field per test after having tested all of them in previous tests. It is because there is no chance that the fields change if we change the protocol or the IP version of the socket. However, in other cases, like for the socket-collector gadget, the fields are extracted from different structs based on the state of the TCP socket, or if they are TCP or UDP. In that case, I think we should test all fields every time. Does it make sense to you?

Reply:
> What would be difficult? You will have the name and value of the field that doesn't contain the expected result.

Yes, you can analyze the output of the tests to get more details. But it would be easier if a single test that says "ip version" is broken instead of having to read all logs to understand the ip version field was broken. 

> Does it make sense to you?

Yes, I think we can use some knowledge about the gadgets to decide how to implement the unit tests for them. We don't have a rule written in stone, if for a given gadget it makes more sense to check all the fields every time that's fine to me.